### PR TITLE
fix: independent components sharing an hf_subdir no longer collide on save

### LINKED ISSUE #621: Chore/fix: `ModelSaver` save-path collision for SeedVR2, then add its tiny test
Another tweak arising from **tiny test** work

---
# PRD: Fix `ModelSaver` save-path collision for SeedVR2, then add its tiny test

## Background

`SeedVR2WeightDefinition3B`/`SeedVR2WeightDefinition7B`
(`src/mflux/models/seedvr2/weights/seedvr2_weight_definition.py`) each define two
components, `transformer` and `vae`, and give **both** the same `hf_subdir="."`:

```python
class SeedVR2WeightDefinition3B:
    @staticmethod
    def get_components() -> List[ComponentDefinition]:
        return [
            ComponentDefinition(
                name="transformer",
                hf_subdir=".",
                num_blocks=32,
                ...
                weight_files=["seedvr2_ema_3b_fp16.safetensors"],
            ),
            ComponentDefinition(
                name="vae",
                hf_subdir=".",
                num_blocks=4,
                ...
                weight_files=["ema_vae_fp16.safetensors"],
            ),
        ]
```

This is correct for **loading**: the real SeedVR2 HF repo has both `.safetensors`
files sitting flat at repo root, and `weight_files` disambiguates them by exact
filename on the load path.

It is **not safe for saving**. `ModelSaver._save_weights`
(`src/mflux/models/common/weights/saving/model_saver.py`) ignores `weight_files`
entirely — it only ever knows about `hf_subdir` (called `subdir` in that function) —
and always writes shard files as plain sequential names:

```python
@staticmethod
def _save_weights(base_path: str, bits: int, model: nn.Module, subdir: str) -> None:
    path = Path(base_path) / subdir
    path.mkdir(parents=True, exist_ok=True)
    weights = dict(tree_flatten(model.parameters()))
    shards = ModelSaver._split_weights(weights)
    for i, shard in enumerate(shards):
        shard_filename = f"{i}.safetensors"
        mx.save_safetensors(str(path / shard_filename), shard, {...})
        ...
    # writes path / "model.safetensors.index.json"
```

With both `transformer` and `vae` resolving `path` to the same directory
(`base_path / "."`), saving the second component overwrites the first's
`0.safetensors`, `1.safetensors`, ... and its `model.safetensors.index.json`.

Confirmed empirically (2026-08-17) via a standalone repro: saving tiny `transformer` +
`vae` components left exactly one `0.safetensors` in the output directory, and
reloading reported `transformer: key sets differ` — the transformer component
actually got the VAE's data back on reload. Also confirmed structurally: SeedVR2 is
the **only** weight definition in the repo where two non-tokenizer components share an
`hf_subdir` — every other model gives each component a distinct subdir (e.g.
`z_image`'s `ZImageWeightDefinition` uses `hf_subdir="vae"` / `"transformer"` /
`"text_encoder"` / `"tokenizer"`; same pattern in `ernie_image`, and everywhere else).

## Why this hasn't broken anything yet

`grep -rn "save_model\|ModelSaver" src/mflux/models/seedvr2/` returns nothing — no
SeedVR2 variant class defines `save_model`, and there is no `mflux-save-*` CLI entry
point wired up for SeedVR2 either. The collision is real but currently **latent**:
nothing in the shipped CLI can trigger `ModelSaver.save_model` with a SeedVR2 weight
definition today. It becomes a silent-corruption bug the moment someone adds save
support for SeedVR2 (e.g. quantized-model saving, matching `mflux-save` for other
models) without independently noticing and fixing this.

This was discovered as a side effect of attempting a tiny model-saving test for
SeedVR2 (see `.claude/notes/tiny-tests.md`, "SeedVR2 — reclassified" section) — the
test's save/reload roundtrip exercises exactly the path that would corrupt a real
save, which read-only dimension analysis alone didn't catch.

## Goal

1. Fix `ModelSaver` so two components sharing an `hf_subdir` (a legitimate load-side
   layout) don't collide on save. Two candidate approaches — pick one:
   - **(a) Prefix shard filenames by component name** in `_save_weights`/
     `_split_weights`, e.g. `f"{component_name}.{i}.safetensors"` instead of bare
     `f"{i}.safetensors"`, and likewise namespace `model.safetensors.index.json`
     per component (or merge weight_maps into one index file per subdir, keyed
     unambiguously). Minimal, doesn't touch `hf_subdir` semantics at all.
   - **(b) Give each component an independent save-time subdirectory**, decoupled
     from `hf_subdir` (which stays as the *load*-side layout hint). E.g. add a
     `save_subdir` field to `ComponentDefinition` (defaulting to `hf_subdir` for
     every other model, where it's already unique) that `ModelSaver` uses instead.
   Check whether either approach has knock-on effects on `WeightLoader`'s
   `_try_load_mflux_format` path (the round-trip re-load of an mflux-saved
   checkpoint) before committing to one — that path currently assumes the
   `hf_subdir`-based layout it wrote.
2. Confirm the fix doesn't change save/load behavior for every other model (all of
   which already have unique per-component `hf_subdir`s, so the fix should be a
   no-op for them — verify, don't just assume).
3. Once (1)-(2) land, write `tests/model_saving/test_tiny_model_saving_seedvr2.py`
   following the `mflux-model-tiny-test` skill. The dimension/shrinkability side was
   already fully worked out and is not blocked — see the specific tiny dimensions
   below.

## Non-goals

- Not adding `mflux-save` / a CLI entry point for SeedVR2 — this PRD only fixes the
  underlying collision so that *whenever* save support is added (here or elsewhere),
  it isn't silently broken. Adding the CLI surface is a separate decision.
- Not touching the FLUX.1/FIBO shared-class parameterization
  (`prd-parameterize-flux-timestep-norm.md`) or Depth Pro
  (`prd-parameterize-depth-pro.md`) — unrelated bugs in unrelated models.
- Not revisiting `weight_files`' role on the load side — it's correct as-is and
  unaffected by this fix.

## Already-solved: tiny-test dimensions (no further investigation needed)

Confirmed 2026-08-17 that every class under `seedvr2/model/` (`SeedVR2Transformer`,
`SeedVR2VAE`, `TransformerBlock`, `MMAttention`, `SwiGLUMLP`, `TimeEmbedding`,
`PatchIn`/`PatchOut`, `ResnetBlock3D`, `Attention3D`, `DownBlock3D`/`UpBlock3D`) takes
explicit dimension kwargs — this part was never blocked. For when the `ModelSaver`
fix lands:

- SeedVR2's real VAE is itself shrinkable (`block_out_channels`, `latent_channels`,
  `layers_per_block`) — use a real tiny `SeedVR2VAE`, `TinyVAEStandIn` isn't needed.
- Use `PatchIn(vid_in_channels=16, ...)`, not the default `33` — `33` produces
  `Linear(132, dim)` and 132 isn't a multiple of 64, so the model's own
  `quantization_predicate` would silently skip quantizing it.
- VAE `block_out_channels` must be a multiple of 32 (hardcoded
  `GroupNorm(num_groups=32)`) *and* 64 (quantization) — `(64, 64)` satisfies both.
- No `quantization_group_size` override (defaults to 64); no `skip_quantization`
  flags (uses `quantization_predicate` instead, which the harness already applies
  symmetrically — not the known krea2/qwen gap).
- Test only `SeedVR2WeightDefinition3B` (or whatever `resolve()` defaults to) — one
  test covers both 3B/7B, since `WeightLoader._load_component` tries
  `_try_load_mflux_format` first and returns on the tiny test's own saved metadata
  before `num_blocks` is ever consulted.

## Verification plan

Unlike the FLUX.1/FIBO and Depth Pro PRDs, this is a **test-infrastructure /
save-path bug fix**, not a change to a model's forward-pass math — so it does not
need the `mflux-debugging` export-then-compare numerical parity workflow. What it
does need:

1. Unit-level: after the fix, save a tiny `transformer` + `vae` pair (or reuse
   whatever repro script surfaced the bug) and confirm the output directory contains
   correctly separated files for both components — no filename collision, both
   reload with matching key sets and values.
2. Confirm no other model's save output changes at all — every other weight
   definition already uses unique `hf_subdir`s, so a diff of saved output
   layout/filenames for e.g. `z_image` or `ernie_image` before/after should be empty.
3. If SeedVR2 save support exists by the time this is picked up, run one real (or
   tiny) save + reload end-to-end and confirm both components come back correctly.
4. `just test-fast` / `just test` green.

## Acceptance criteria

- [ ] `ModelSaver` (or `ComponentDefinition`) changed so components sharing an
      `hf_subdir` no longer collide on save.
- [ ] Confirmed no behavior change for models with already-unique `hf_subdir`s.
- [ ] `tests/model_saving/test_tiny_model_saving_seedvr2.py` written and green, using
      the dimensions already worked out above.
- [ ] `.claude/notes/tiny-tests.md` coverage matrix updated: `seedvr2` from "❌
      blocked" to "✅".
- [ ] Consider whether `WeightLoader`'s `_try_load_mflux_format` needs a matching
      change, or whether it already tolerates the new save layout unchanged (it reads
      `model.safetensors.index.json` and `weight_map`, which any of the fix
      approaches above should keep intact — verify, don't assume).

## References

- `.claude/notes/tiny-tests.md` — "SeedVR2 — reclassified: blocked by a real
  ModelSaver bug, not dimensions" section (full investigation + repro details).
- `.cursor/skills/mflux-model-tiny-test/SKILL.md` — tiny test pattern.
- `src/mflux/models/common/weights/saving/model_saver.py` — `_save_weights`, the
  function that needs to change.
- `src/mflux/models/seedvr2/weights/seedvr2_weight_definition.py` — both
  `SeedVR2WeightDefinition3B` and `...7B` need their `ComponentDefinition`s to keep
  working after the fix (or gain a new field, per approach (b) above).
- `src/mflux/models/common/weights/loading/weight_definition.py` — `ComponentDefinition`
  dataclass, if approach (b) (a new `save_subdir` field) is chosen.
- Sibling PRDs for the same "investigation said trivial, implementation found a real
  bug" pattern: `prd-parameterize-flux-timestep-norm.md` (FLUX.1/FIBO),
  `prd-parameterize-depth-pro.md` (Depth Pro) — different root causes, same overall
  shape of "someone needs to scope and verify a `src/` fix before the tiny test can
  land."

### PR DESCRIPTION
**Pull request by IonDen**
## What

`ModelSaver` corrupts a saved checkpoint when two independent components share an `hf_subdir`. SeedVR2's `transformer` and `vae` both declare `hf_subdir="."`, which is right for loading the real repo, where `weight_files` tells them apart. On save, `_save_weights` ignores `weight_files` and writes each component's shards as `0.safetensors`, `1.safetensors`, ... plus `model.safetensors.index.json`, so saving the second component overwrites the first's shards and index. Reloading then hands each component the other's weights.

The fix gives each independent component that shares an `hf_subdir` its own `<hf_subdir>/<name>` directory, used by both the saver and the mflux-format reload probe. The HuggingFace fallback still reads `hf_subdir`, so loading a real (never-mflux-saved) repo is unchanged. Components that split one shared source by prefix, like FIBO VLM's decoder and visual, keep sharing their directory, and every model with unique subdirs is untouched. A scan of all weight definitions confirms only SeedVR2 is affected.

The bug is latent today, since no SeedVR2 save path is wired up, so this changes nothing users can currently trigger and guards the collision for whenever SeedVR2 save support lands.

Fixes #621.

## Verification

Two tests, both no-weights and in the default selector, saving and reloading through the real ModelSaver, WeightLoader and WeightApplier and asserting each component's parameter tree survives byte for byte. `test_tiny_model_saving_seedvr2.py` builds real SeedVR2 transformer and VAE classes at tiny dimensions, the case the bug is about. `test_model_saver_shared_subdir.py` is a minimal two-component stand-in that isolates the same collision. Both are red without the fix, where the transformer reloads with the VAE's weights, and green with it.

Fast model-saving suite 154 passed, weight-loading and resolution 226 passed, ruff clean on the tree.

## Release note

```release-note
none
```

## Summary by CodeRabbit

- **Bug Fixes**
  - Fixed checkpoint saving and loading for models whose components share a source directory.
  - Ensured components are stored and restored from distinct locations when needed.
  - Prevented component files from overwriting one another in shared model layouts.
  - Added validation to report conflicting save locations instead of silently overwriting files.
  - Preserved compatibility with existing model formats and layouts.

- **Tests**
  - Added coverage for shared-directory and quantized checkpoint round trips, including exact weight preservation.

<h3>Greptile Summary</h3>

The PR prevents independently saved components with the same Hugging Face source directory from overwriting one another and aligns mflux-format reload probing with the new save layout.
- Adds deterministic per-component save directories for independent shared-subdirectory components.
- Preserves shared-source prefix-filtered layouts and the original Hugging Face fallback paths.
- Adds generic and SeedVR2-specific quantized checkpoint round-trip coverage.

<h3>Confidence Score: 5/5</h3>

The PR appears safe to merge, with saver and loader directory selection aligned and focused round-trip coverage for the affected layout.

Independent components sharing an hf_subdir are assigned matching noncolliding save and reload paths, while existing source-checkpoint fallback behavior and intentional prefix-filtered sharing remain intact.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| src/mflux/models/common/weights/loading/weight_definition.py | Centralizes save-directory selection while preserving intentional prefix-filtered shared-source layouts. |
| src/mflux/models/common/weights/loading/weight_loader.py | Probes the computed component save directory for mflux checkpoints while retaining the original Hugging Face fallback. |
| src/mflux/models/common/weights/saving/model_saver.py | Writes independently saved shared-subdirectory components into distinct directories to prevent shard and index replacement. |
| tests/model_saving/test_model_saver_shared_subdir.py | Verifies an isolated quantized two-component shared-directory checkpoint round trip. |
| tests/model_saving/test_tiny_model_saving_seedvr2.py | Exercises the corrected layout through real tiny SeedVR2 transformer and VAE components. |

</details>

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart LR
  D[Component definitions] --> S[Compute save subdirectories]
  S -->|Unique source directory| H[Use hf_subdir]
  S -->|Shared and independent| N[Use hf_subdir/name]
  S -->|Shared prefix-filtered source| H
  N --> W[Save mflux shards and index]
  H --> W
  W --> L[Reload model]
  L --> P[Probe computed mflux save directory]
  P -->|Not found| F[Fall back to original Hugging Face directory]
```
</details>

<sub>Reviews (1): Last reviewed commit: ["fix: independent components sharing an h..."](https://github.com/mflux-community/mflux/commit/413e1e12100d71436a591d7684b1bcca05c64676) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=60980886)</sub>

**Context used:**

- Knowledge Base — [Model loading and adaptation](https://app.greptile.com/mflux-community/-/custom-context/knowledge-base/mflux-community/mflux/-/docs/model-loading-and-adaptation.md)

### CODE REVIEW

File: src/mflux/models/common/weights/loading/weight_definition.py [hunk 1]
```diff
@@ -52,6 +52,38 @@ class ComponentDefinition:
...
+        # model spells one directory two ways, and normalizing "" to "." would wrongly merge
+        # FIBO VLM's decoder/visual ("") with its fibo_vlm (".") component.
+        counts: dict = {}
+        for c in components:
+            counts[c.hf_subdir] = counts.get(c.hf_subdir, 0) + 1
+        return {
+            c.name: (
+                str(Path(c.hf_subdir) / c.name)
+                if counts[c.hf_subdir] > 1 and c.weight_prefix_filters is None
+                else c.hf_subdir
```
[the remark this answers is not in the document]
Comment (IonDen) [reply src/mflux/models/common/weights/loading/weight_definition.py]: Traced this. The exception is deliberate. FIBO VLM's `decoder` and `visual` are both prefix-filtered and share `hf_subdir=""` on purpose: they read the same files and split them by prefix, and `visual` is nested inside `decoder`, so only `decoder` is ever written. Removing the exception (the suggested diff) namespaces them, so on reload the probe looks in `root/visual`, which was never saved, and `test_tiny_model_saving_fibo_vlm` fails with `FileNotFoundError`.

You're right that a shared directory should not be able to silently overwrite, though. I added a guard in `ModelSaver.save_model` that raises if two components that are actually written ever resolve to the same directory, so the case you describe fails with a clear error instead of corrupting a checkpoint, with a test for it. Also applied the type-annotation nit. Thanks for the catch.

File: src/mflux/models/common/weights/saving/model_saver.py [hunk 2]
```diff
@@ -26,13 +27,38 @@ def save_model(
...
+        # pair is a real attribute, so only one is written. Checked before anything reaches
+        # disk, and by attribute presence so a nested/aliased component that is not written
+        # is not counted.
+        written_dirs: dict[str, str] = {}
+        for component_def in weight_definition.get_components():
+            attr_name = component_def.model_attr or component_def.name
+            if getattr(model, attr_name, None) is None:
+                continue
+            subdir = save_subdirs[component_def.name]
+            if subdir in written_dirs:
```
[the remark this answers is not in the document]
Comment (IonDen) [reply src/mflux/models/common/weights/saving/model_saver.py]: Good catch. The guard now keys on the resolved path, so `""` and `"."` count as one directory and a pair spelled differently is still caught. `save_subdirs` groups the same way, so two independent components spelled `""` and `"."` are separated rather than left to collide. Added tests for both, including the guard case.

**Review by fxd0h — APPROVED** [review]
Reproduced #621 with the real 3B weights before reading the fix. On main, saving SeedVR2 q8 leaves two shards and one index flat at the root; the reload doesn't fail, and the upscale of a 256 px test image comes out with a mean difference of 26/255 against the run from the HF layout. Silent garbage.

With this branch the save lands in `transformer/` and `vae/`, the reload is bit identical (max diff 0), and a checkpoint written by main now fails loudly (`FileNotFoundError` on `seedvr2_ema_3b_fp16.safetensors`) instead of loading. Nothing under mflux-community on HF uses the old flat layout, so the layout change has no victims. Fast suite: 1599 passed.

One pre-existing thing, not for this PR: the "contains no files matching" warning still prints for an mflux-saved SeedVR2 checkpoint, because the `weight_files` probe runs before the mflux-format probe.
